### PR TITLE
Qa/kibana integration

### PR DESCRIPTION
**Pull request by npliashechnikov**
Implement support for the Elastic Search in the test framework

### CODE REVIEW

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+     * @param level - log level (can be null)
+     * @param timeRange - search depth (in seconds, set to 0 to search all).
+     * @param resultCount - max number of returned documents (100 default)
+     * @param defaultField - field in log entry to search at (_source by default)
+     * @param index - Elastic Search index to lookup (_all by default)
+     *
+     * In case you need to lookup multiple application IDs, tags or log levels, pass parameters as:
+     * "VALUE1 OR VALUE2 OR ... OR VALUE_N"
+     */
+    def getLogs(String app_id=null, String tags=null, String level="INFO OR WARN OR ERROR", Map<String,
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy]: Too many input params. Can we take a certain object with required data as input? I would rather do `getLogs(new LogsRequest(level: "INFO", timeRange: 30))`

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                        "default_field": defaultField,
+                        "query": queryString
+                ]
+        ]]
+
+        def jsonBuilder = new JsonBuilder(query)
+
+
+        String uri = "/${index}/_search/?size=${resultCount}"
+        log.info("Issuing elastic query: ${jsonBuilder.toString()}")
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy]: Do you mind having this on debug level?

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+import groovy.json.JsonBuilder
+import org.springframework.stereotype.Component
+
+/**
+ * A set of helper methods to access, read and filter log entries from the log management system
+ */
+
+@Component
+@Slf4j
+class KibanaHelper {
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy]: Did you investigate official clients?
https://www.elastic.co/guide/en/elasticsearch/client/java-api/current/java-search.html
https://www.elastic.co/blog/found-java-clients-for-elasticsearch
https://www.elastic.co/guide/en/elasticsearch/client/java-api/current/java-query-dsl.html

If they are usable, we should be definitely using them.

File: services/src/functional-tests/kilda.properties.example [hunk 4]
```diff
@@ -16,6 +16,10 @@ mininet.endpoint=http://localhost:38080
 mininet-flowtool.endpoint=http://localhost:17191
 spring.profiles.active=virtual
 
+elasticsearch.endpoint = https://elasticsearch.local:9200
+elasticsearch.username = username
+elasticsearch.password = password
+
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/kilda.properties.example]: Will do, thanks.

File: services/src/test-library/src/main/java/org/openkilda/testing/service/elastic/ElasticHelperImpl.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    @Qualifier("elasticSearchRestTemplate")
+    private RestTemplate restTemplate;
+
+    /**
+     * Inquiry ElasticSearch for logs. Not intended to be used directly.
+     * @param uri - Search address and parame
+     * @param query - Search query (with headers and ES/Lucene JSON query body)
+     * @return HTTP response from ES
+     */
+    public HashMap getLogs(String uri, HttpEntity query) {
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/test-library/src/main/java/org/openkilda/testing/service/elastic/ElasticHelperImpl.java]: I originally meant to keep Java portion as generic as possible, allowing the API user to design their own query interfaces as needed, depending on the tasks. So I'd rather leave the code as is for the moment.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy (hunk 1 above)
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy]: Will do that.

File: services/src/test-library/src/main/java/org/openkilda/testing/service/elastic/ElasticHelperImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (rtretyak) [reply services/src/test-library/src/main/java/org/openkilda/testing/service/elastic/ElasticHelperImpl.java]: Ok, then this should be part of the _query_, not _uri_. We should not specify _uri_ from tests

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy (hunk 1 above)
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy]: Added a builder-like interface that accepts Closure to add all required parameters.
Also, added a simplified interface that omits most of the unnecessary data.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy (hunk 2 above)
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy]: Will do, thanks.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy (hunk 3 above)
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/KibanaHelper.groovy]: Unfortunately, wiring up a full-fledged elastic client turned out to be way too complicated for our simple tasks (and it'll require users to get into ES details, which are not pretty)

**Comment by rtretyak** [comment]
not relevant, see #1433

**Pull request closed without merging**